File: frontend/lib/services/auth_service.dart
import '../models/user_model.dart';
import 'api_service.dart';
import 'social/google_login_service.dart';
import 'social/kakao_login_service.dart';
import 'social/naver_login_service.dart';
import 'token_storage.dart';

/// 회원(인증) 도메인 서비스.
/// 회원가입/로그인/로그아웃/회원탈퇴/SNS 간편로그인/토큰 재발급을 담당한다.
class AuthService {
  final ApiService _apiService;
  final TokenStorage _tokenStorage;
  final GoogleLoginService _googleLoginService;
  final KakaoLoginService _kakaoLoginService;
  final NaverLoginService _naverLoginService;

  AuthService({
    ApiService? apiService,
    TokenStorage? tokenStorage,
    GoogleLoginService? googleLoginService,
    KakaoLoginService? kakaoLoginService,
    NaverLoginService? naverLoginService,
  })  : _apiService = apiService ?? ApiService(),
        _tokenStorage = tokenStorage ?? TokenStorage(),
        _googleLoginService = googleLoginService ?? GoogleLoginService(),
        _kakaoLoginService = kakaoLoginService ?? KakaoLoginService(),
        _naverLoginService = naverLoginService ?? NaverLoginService();

  /// 일반 회원가입.
  Future<void> signup({required String email, required String password, required String name}) async {
    await _apiService.post('/auth/signup', {
      'email': email,
      'password': password,
      'name': name,
    });
  }

  /// 일반(이메일/비밀번호) 로그인. 성공 시 토큰을 보안 저장소에 저장한다.
  Future<UserModel> login({required String email, required String password}) async {
    final data = await _apiService.post('/auth/login', {'email': email, 'password': password});
    return _handleAuthResult(data as Map<String, dynamic>);
  }

  /// SNS 간편로그인. [provider]에 맞는 SDK로 먼저 로그인한 뒤,
  /// 발급받은 Access Token을 백엔드로 전달해 자체 로그인(JWT 발급)을 완료한다.
  /// 사용자가 SNS 로그인을 취소하면 null을 반환한다.
  Future<UserModel?> socialLogin(AuthProvider provider) async {
    final providerAccessToken = await _fetchSocialAccessToken(provider);
    if (providerAccessToken == null) return null;

    final data = await _apiService.post('/auth/social/${provider.name}', {
      'accessToken': providerAccessToken,
    });
    return _handleAuthResult(data as Map<String, dynamic>);
  }

  /// provider별 SDK를 호출해 Access Token을 가져오는 내부 함수.
  Future<String?> _fetchSocialAccessToken(AuthProvider provider) {
    switch (provider) {
      case AuthProvider.kakao:
        return _kakaoLoginService.signIn();
      case AuthProvider.naver:
        return _naverLoginService.signIn();
      case AuthProvider.google:
        return _googleLoginService.signIn();
      case AuthProvider.local:
        throw ArgumentError('local은 SNS 로그인 provider가 아닙니다.');
    }
  }

  /// 로그인/간편로그인 응답(JSON)에서 토큰을 저장하고 사용자 정보를 반환하는 공통 처리 함수.
  Future<UserModel> _handleAuthResult(Map<String, dynamic> data) async {
    final accessToken = data['accessToken'] as String;
    final refreshToken = data['refreshToken'] as String;
    await _tokenStorage.saveTokens(accessToken: accessToken, refreshToken: refreshToken);
    return UserModel.fromJson(data['user'] as Map<String, dynamic>);
  }

  /// 저장된 Refresh Token으로 Access Token을 재발급받아 갱신한다.
  Future<void> refreshAccessToken() async {
    final refreshToken = await _tokenStorage.readRefreshToken();
    if (refreshToken == null) {
      throw StateError('저장된 refreshToken이 없습니다. 다시 로그인해야 합니다.');
    }

    final data = await _apiService.post('/auth/refresh', {'refreshToken': refreshToken});
    await _tokenStorage.updateAccessToken((data as Map<String, dynamic>)['accessToken'] as String);
  }

  /// 로그아웃 처리. 서버의 Refresh Token을 무효화하고 로컬 토큰을 삭제한다.
  Future<void> logout() async {
    await _apiService.post('/auth/logout', {}, authorized: true);
    await _tokenStorage.clear();
  }

  /// 회원 탈퇴 처리. 서버에 탈퇴를 요청하고 로컬 토큰을 삭제한다.
  Future<void> withdraw() async {
    await _apiService.delete('/auth/withdraw', authorized: true);
    await _tokenStorage.clear();
  }

  /// 로그인 여부(Access Token 보유 여부)를 확인한다. 스플래시 화면의 자동 로그인 분기에 사용한다.
  Future<bool> isLoggedIn() async {
    final accessToken = await _tokenStorage.readAccessToken();
    return accessToken != null;
  }
}

File: backend/src/services/auth.service.js
const userService = require('./user.service');
const tokenService = require('./token.service');
const oauthService = require('./oauth.service');
const { hash, compare } = require('../utils/hash');
const AppError = require('../utils/AppError');

/**
 * Access/Refresh Token 쌍을 발급하고, Refresh Token은 해시로 DB에 저장한다.
 * (탈취 대비를 위해 평문 그대로 저장하지 않는다)
 * @param {import('../models/user.model')} user
 * @returns {Promise<{accessToken: string, refreshToken: string}>}
 */
async function issueTokenPair(user) {
  const accessToken = tokenService.generateAccessToken(user);
  const refreshToken = tokenService.generateRefreshToken(user);

  const hashedRefreshToken = await hash(refreshToken);
  await userService.updateRefreshToken(user, hashedRefreshToken);

  return { accessToken, refreshToken };
}

/**
 * 일반 회원가입 처리.
 * @param {{email: string, password: string, name: string}} payload
 * @returns {Promise<import('../models/user.model')>} 생성된 사용자
 */
async function signup({ email, password, name }) {
  const existingUser = await userService.findByEmailAndProvider(email, 'local');
  if (existingUser) {
    throw new AppError(409, '이미 가입된 이메일입니다.');
  }

  const hashedPassword = await hash(password);
  return userService.createLocalUser({ email, hashedPassword, name });
}

/**
 * 일반 로그인 처리. 이메일/비밀번호 검증 후 토큰을 발급한다.
 * @param {{email: string, password: string}} payload
 * @returns {Promise<{accessToken: string, refreshToken: string, user: import('../models/user.model')}>}
 */
async function login({ email, password }) {
  const user = await userService.findByEmailAndProvider(email, 'local');

  // 사용자가 없거나 이미 탈퇴한 계정이면 동일한 에러 메시지로 응답한다. (계정 존재 여부 노출 방지)
  if (!user || user.status === 'withdrawn') {
    throw new AppError(401, '이메일 또는 비밀번호가 올바르지 않습니다.');
  }

  const isPasswordMatched = await compare(password, user.password);
  if (!isPasswordMatched) {
    throw new AppError(401, '이메일 또는 비밀번호가 올바르지 않습니다.');
  }

  const tokens = await issueTokenPair(user);
  return { ...tokens, user };
}

/**
 * SNS 간편로그인 처리.
 * Flutter 앱이 각 SNS SDK(kakao_flutter_sdk, flutter_naver_login, google_sign_in)로
 * 먼저 로그인해 발급받은 accessToken을 그대로 전달받아, 해당 SNS API로 프로필을 조회한다.
 * 최초 로그인이면 회원을 자동 생성(가입)하고, 기존 회원이면 그대로 로그인 처리한다.
 * @param {string} provider - 'naver' | 'kakao' | 'google'
 * @param {string} providerAccessToken - SNS 제공자로부터 발급받은 Access Token
 * @returns {Promise<{accessToken: string, refreshToken: string, user: import('../models/user.model')}>}
 */
async function socialLogin(provider, providerAccessToken) {
  const profile = await oauthService.getSocialProfile(provider, providerAccessToken);

  let user = await userService.findByProviderId(provider, profile.providerId);

  if (!user) {
    // 최초 SNS 로그인 -> 자동 회원가입.
    // SNS에서 이메일 제공에 동의하지 않은 경우를 대비해 대체 이메일을 생성한다.
    const email = profile.email ?? `${provider}_${profile.providerId}@social.project-heritage`;
    user = await userService.createSocialUser({
      provider,
      providerId: profile.providerId,
      email,
      name: profile.name,
    });
  }

  if (user.status === 'withdrawn') {
    throw new AppError(403, '탈퇴한 계정입니다.');
  }

  const tokens = await issueTokenPair(user);
  return { ...tokens, user };
}

/**
 * Refresh Token으로 새로운 Access Token을 재발급한다.
 * @param {string} refreshToken - 클라이언트가 보관 중인 Refresh Token
 * @returns {Promise<{accessToken: string}>}
 */
async function refreshAccessToken(refreshToken) {
  let payload;
  try {
    payload = tokenService.verifyRefreshToken(refreshToken);
  } catch (error) {
    throw new AppError(401, '리프레시 토큰이 유효하지 않거나 만료되었습니다.');
  }

  const user = await userService.findById(payload.id);
  if (!user || !user.refreshToken) {
    throw new AppError(401, '다시 로그인해 주세요.');
  }

  // DB에 저장된 해시와 비교해, 로그아웃되었거나 탈취된 토큰이 아닌지 확인한다.
  const isTokenMatched = await compare(refreshToken, user.refreshToken);
  if (!isTokenMatched) {
    throw new AppError(401, '리프레시 토큰이 일치하지 않습니다.');
  }

  const accessToken = tokenService.generateAccessToken(user);
  return { accessToken };
}

/**
 * 로그아웃 처리. DB에 저장된 Refresh Token을 제거해 재사용을 막는다.
 * @param {number} userId - 로그인 중인 사용자 ID (인증 미들웨어에서 추출)
 * @returns {Promise<void>}
 */
async function logout(userId) {
  const user = await userService.findById(userId);
  if (!user) return;
  await userService.updateRefreshToken(user, null);
}

/**
 * 회원 탈퇴 처리. 계정을 소프트 삭제(status=withdrawn) 하고 민감 정보를 제거한다.
 * @param {number} userId - 로그인 중인 사용자 ID (인증 미들웨어에서 추출)
 * @returns {Promise<void>}
 */
async function withdraw(userId) {
  const user = await userService.findById(userId);
  if (!user) {
    throw new AppError(404, '사용자를 찾을 수 없습니다.');
  }
  await userService.markAsWithdrawn(user);
}

module.exports = { signup, login, socialLogin, refreshAccessToken, logout, withdraw };
